repair/repair.cc

```diff
@@ -1191,12 +1194,18 @@ int repair_service::do_repair_start(sstring keyspace, std::unordered_map<sstring
             throw std::runtime_error("aborted by user request");
         }
 
+        auto ring_version = erm->get_token_metadata().get_ring_version();
         for (auto shard : boost::irange(unsigned(0), smp::count)) {
             auto f = container().invoke_on(shard, [keyspace, table_ids, id, ranges, hints_batchlog_flushed,
-                    data_centers = options.data_centers, hosts = options.hosts, ignore_nodes] (repair_service& local_repair) mutable {
+                    data_centers = options.data_centers, hosts = options.hosts, ignore_nodes, ring_version] (repair_service& local_repair) mutable {
+                auto erm = local_repair.get_db().local().find_keyspace(keyspace).get_effective_replication_map();
+                auto local_ring_version = erm->get_token_metadata().get_ring_version();
+                if (local_ring_version != ring_version) {
```

Comment:
This is good for now (and an improvement) but probably insufficient. For example, we can get a different erm if we have replication strategy parameter changes.

Reply:
It's a bit subtle but the schema change is processed by storage_service::on_update_keyspace that calculates a new e_r_m and when applying it we recalculate the token_metadata, although it's didn't explicitly change, and give it a new version, so nowadays the token_metadata::ring_version covers both the token_metadata and the effective_replication_map.  Topology changes should also cause recalculation of this bundle.
Maybe we need to define a `global_effective_replication_map` that will get the e_r_m on all shards (with grabbing the shared_token_metadata lock so it won't change) and then we can pass it around across shards.